python/dgl/heterograph.py

```diff
@@ -0,0 +1,1387 @@
+
+        Game name   Tetris  Minecraft
+        Game ID     0       1
+
+        Developer name  Nintendo    Mojang
+        Developer ID    0           1
+
+    One can construct the graph as follows:
+
+    >>> import networkx as nx
+    >>> metagraph = nx.MultiGraph([
+    ...     ('user', 'user', 'follows'),
+    ...     ('user', 'game', 'plays'),
+    ...     ('developer', 'game', 'develops')])
+    >>> g = DGLBaseHeteroGraph(
+    ...     metagraph=metagraph,
+    ...     number_of_nodes_by_type={'user': 4, 'game': 2, 'developer': 2},
+    ...     edge_connections_by_type={
+    ...         # Alice follows Bob and Bob follows Carol
+    ...         ('user', 'user', 'follows'): ([0, 1], [1, 2]),
+    ...         # Alice and Bob play Tetris and Bob and Carol play Minecraft
+    ...         ('user', 'game', 'plays'): ([0, 1, 1, 2], [0, 0, 1, 1]),
+    ...         # Nintendo develops Tetris and Mojang develops Minecraft
+    ...         ('developer', 'game', 'develops'): ([0, 1], [0, 1])})
+    """
+
+    # pylint: disable=unused-argument
+    def __init__(
+            self,
+            metagraph,
+            number_of_nodes_by_type,
+            edge_connections_by_type):
+        super(DGLBaseHeteroGraph, self).__init__()
+
+    def __getitem__(self, key):
+        """Returns a view on the heterogeneous graph with given node/edge
+        type:
+
+        * If ``key`` is a str, it returns a heterogeneous subgraph induced
+          from nodes of type ``key``.
+        * If ``key`` is a pair of str (type_A, type_B), it returns a
+          heterogeneous subgraph induced from the union of both node types.
+        * If ``key`` is a triplet of str
+
+              (src_type_name, dst_type_name, edge_type_name)
+
+          It returns a heterogeneous subgraph induced from the edges with
+          source type name ``src_type_name``, destination type name
+          ``dst_type_name``, and edge type name ``edge_type_name``.
+
+        The view would share the frames with the parent graph; any
+        modifications on one's frames would reflect on the other.
+
+        Note that the subgraph itself is not materialized until someone
+        queries the subgraph structure.  This implies that calling computation
+        methods such as
+
+            g['user'].update_all(...)
+
+        would not create a subgraph of users.
+
+        Parameters
+        ----------
+        key : str or tuple
+            See above
+
+        Returns
+        -------
+        DGLBaseHeteroGraphView
+            The induced subgraph view.
+        """
+        pass
+
+
+class DGLBaseHeteroGraphView(object):
+    """View on a heterogeneous graph, constructed from
+    DGLBaseHeteroGraph.__getitem__().
+
+    It is semantically the same as a subgraph, except that
+
+    * The subgraph itself is not materialized until the user explicitly
+    queries the subgraph structure (e.g. calling ``in_edges``, but not
+    ``update_all``).
+    """
+    pass
```

Comment:
what are the methods in this class?

Reply:
The methods in `DGLBaseGraph` *would* materialize the subgraph since they explicitly ask for the structure of the subgraph.
For message passing APIs we clearly don't have to materialize it.

Reply:
If the class is inherited from `DGLBaseHeteroGraph`, there won't be computation APIs. It should be inherited from `DGLHeteroGraph`?

Reply:
I don't think we should materialize the subgraph for query APIs. For example, `g['user'].number_of_nodes()`. If we need to materialize the subgraph before getting the number of nodes. It'll be too expensive.